File: texter/src/main/java/pl/org/seva/texter/activity/HomeLocationActivity.java
/*
 * Copyright (C) 2016 Wiktor Nizio
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package pl.org.seva.texter.activity;

import android.Manifest;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import pl.org.seva.texter.R;
import pl.org.seva.texter.application.TexterApplication;
import pl.org.seva.texter.dagger.Graph;
import pl.org.seva.texter.databinding.ActivityHomeLocationBinding;
import pl.org.seva.texter.manager.GpsManager;
import pl.org.seva.texter.manager.LastLocationManager;
import pl.org.seva.texter.manager.PermissionsManager;
import pl.org.seva.texter.utils.Constants;

public class HomeLocationActivity extends AppCompatActivity implements
        GoogleMap.OnMapLongClickListener,
        GoogleMap.OnCameraIdleListener {

    private GpsManager gpsManager;
    private LastLocationManager lastLocationManager;
    private PermissionsManager permissionsManager;

    private static final String STATE = "STATE";
    private static final String ZOOM_PROPERTY_NAME = "map_preference_gui_zoom";
    private static final float ZOOM_DEFAULT_VALUE = 7.5f;

    private Disposable locationChangedSubscription = Disposables.empty();

    /** Latitude. */
    private double lat;
    /** Longitude. */
    private double lon;
    private boolean toastShown;
    private float zoom;
    private boolean isCurrentLocationAvailable;

    private GoogleMap map;
    private Button useCurrentButton;
    /** True indicates restoring from a saved state. */
    private boolean animateCamera = true;
    private int mapContainerId;
    private MapFragment mapFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            HomeLocationActivity.SavedState myState =
                    (HomeLocationActivity.SavedState) savedInstanceState.get(STATE);
            lat = myState.lat;
            lon = myState.lon;
            toastShown = myState.toastShown;
            zoom = myState.zoom;
            animateCamera = false;
        }

        Graph graph = ((TexterApplication) getApplication()).getGraph();
        gpsManager = graph.gpsManager();
        lastLocationManager = graph.lastLocationManager();
        permissionsManager = graph.permissionsManager();

        String value = getPersistedString();
        lat = parseLatitude(value);
        lon = parseLongitude(value);

        ActivityHomeLocationBinding binding =
                DataBindingUtil.setContentView(this, R.layout.activity_home_location);
        mapContainerId = binding.mapContainer.getId();
        MapsInitializer.initialize(this);
        useCurrentButton = binding.currentLocationButton;

        boolean locationPermitted = ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION) ==
                PackageManager.PERMISSION_GRANTED;
        boolean locationAvailable = locationPermitted &&
                lastLocationManager.isLocationAvailable();
        useCurrentButton.setEnabled(locationAvailable);
        if (!toastShown) {
            Toast.makeText(
                    this,
                    R.string.long_press,
                    Toast.LENGTH_SHORT).show();
            toastShown = true;
        }
        locationChangedSubscription = gpsManager.locationChangedListener().subscribe(
                __ -> onLocationChanged());

        if (!locationPermitted) {
            useCurrentButton.setEnabled(false);
            setLocationPermissionListeners();
        }
    }

    private void setLocationPermissionListeners() {
        permissionsManager
                .permissionGrantedListener()
                .filter(permission -> permission.equals(Manifest.permission.ACCESS_FINE_LOCATION))
                .subscribe(__ -> onLocationPermissionGranted());
    }

    @Override
    protected void onResume() {
        super.onResume();
        FragmentManager fm = getFragmentManager();
        mapFragment = (MapFragment) fm.findFragmentByTag("map");
        if (mapFragment == null) {
            mapFragment = new MapFragment();
            fm.beginTransaction().add(mapContainerId, mapFragment, "map").commit();
        }
        mapFragment.getMapAsync(googleMap -> {
            map = googleMap;
            if (ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.ACCESS_FINE_LOCATION) ==
                    PackageManager.PERMISSION_GRANTED) {
                map.setMyLocationEnabled(true);
            }
            else {
                setLocationPermissionListeners();
            }
            zoom = PreferenceManager.getDefaultSharedPreferences(this).
                    getFloat(ZOOM_PROPERTY_NAME, ZOOM_DEFAULT_VALUE);

            updateMarker();
            CameraPosition cameraPosition = new CameraPosition.Builder()
                    .target(new LatLng(lat, lon)).zoom(zoom).build();
            if (animateCamera) {
                map.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
            }
            else {
                map.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
            }
            animateCamera = false;

            map.setOnMapLongClickListener(this);
            map.setOnCameraIdleListener(this);
        });
    }

    @Override
    public void onBackPressed() {
        locationChangedSubscription.dispose();
        toastShown = false;

        persistString(toString());
        PreferenceManager.getDefaultSharedPreferences(this).edit().
                putFloat(ZOOM_PROPERTY_NAME, zoom).apply();
        gpsManager.onHomeLocationChanged();

        if (mapFragment != null) {
            // Without enclosing in the if, throws:
            // java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
            getFragmentManager().beginTransaction().remove(mapFragment).commit();
        }
        super.onBackPressed();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        final HomeLocationActivity.SavedState myState
                = new HomeLocationActivity.SavedState();
        myState.lat = lat;
        myState.lon = lon;
        myState.toastShown = toastShown;
        myState.zoom = zoom;

        if (mapFragment != null) {
            getFragmentManager().beginTransaction().remove(mapFragment).commit();
            mapFragment = null;
        }
        super.onSaveInstanceState(outState);
    }

    public String toString() {
        return toString(lat, lon);
    }

    private static String toString(double lat, double lon) {
        return ("geo:") +
                (int) lat + "." +
                Double.toString(lat - (int) lat).substring(2, 8) + "," +
                (int) lon + "." +
                Double.toString(lon - (int) lon).substring(2, 8);
    }

    private static double parseLatitude(String uri) {
        String str = uri.substring(uri.indexOf(":") + 1, uri.indexOf(","));
        return Double.valueOf(str);
    }

    private static double parseLongitude(String uri) {
        String str = uri.substring(uri.indexOf(",") + 1);
        return Double.valueOf(str);
    }

    private void persistString(String val) {
        PreferenceManager.getDefaultSharedPreferences(this).edit()
            .putString(SettingsActivity.HOME_LOCATION, val).apply();
    }

    private String getPersistedString() {
        return PreferenceManager.getDefaultSharedPreferences(this).
                getString(SettingsActivity.HOME_LOCATION, Constants.DEFAULT_HOME_LOCATION);
    }

    private void updateMarker() {
        map.clear();
        MarkerOptions marker = new MarkerOptions().position(
                new LatLng(lat, lon)).
                title(getString(R.string.home));
        marker.icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_ROSE));
        map.addMarker(marker);
    }

    private void onLocationChanged() {
        if (map == null || useCurrentButton == null) {
            return;
        }
        isCurrentLocationAvailable = true;
        useCurrentButton.setEnabled(true);
    }

    private void onLocationPermissionGranted() {
        if (map != null) {
            //noinspection MissingPermission
            map.setMyLocationEnabled(true);
            useCurrentButton.setEnabled(false);
        }
    }

    @Override
    public void onMapLongClick(LatLng latLng) {
        lat = latLng.latitude;
        lon = latLng.longitude;
        updateMarker();
        if (isCurrentLocationAvailable) {
            useCurrentButton.setEnabled(true);
        }
    }

    @Override
    public void onCameraIdle() {
        zoom = map.getCameraPosition().zoom;
    }

    public void onUseCurrentLocationButtonClicked(@SuppressWarnings("UnusedParameters") View view) {
        useCurrentButton.setEnabled(false);
        LatLng loc = lastLocationManager.getLatLng();
        if (loc != null) {
            lat = loc.latitude;
            lon = loc.longitude;
            updateMarker();
            CameraPosition cameraPosition = new CameraPosition.Builder()
                    .target(new LatLng(lat, lon)).
                            zoom(zoom).build();
            map.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
        }
    }


    private static class SavedState implements Parcelable {
        private double lat;
        private double lon;
        private boolean toastShown;
        private float zoom;

        SavedState() {
        }

        SavedState(Parcel source) {
            lat = source.readDouble();
            lon = source.readDouble();
            toastShown = source.readInt() != 0;
            zoom = source.readFloat();
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(@NonNull Parcel dest, int flags) {
            dest.writeDouble(lat);
            dest.writeDouble(lon);
            dest.writeInt(toastShown ? 1 : 0);
            dest.writeFloat(zoom);
        }

        public static final Parcelable.Creator<HomeLocationActivity.SavedState> CREATOR =
            new Parcelable.Creator<HomeLocationActivity.SavedState>() {

                public HomeLocationActivity.SavedState createFromParcel(Parcel in) {
                    return new HomeLocationActivity.SavedState(in);
                }

                public HomeLocationActivity.SavedState[] newArray(int size) {
                    return new HomeLocationActivity.SavedState[size];
                }
            };
    }
}

File: texter/src/main/java/pl/org/seva/texter/view/activity/HomeLocationActivity.kt
/*
 * Copyright (C) 2017 Wiktor Nizio
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package pl.org.seva.texter.view.activity

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.preference.PreferenceManager
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.widget.Button
import android.widget.Toast
import com.google.android.gms.maps.*

import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions

import javax.inject.Inject

import io.reactivex.disposables.Disposables
import pl.org.seva.texter.R
import pl.org.seva.texter.TexterApplication
import pl.org.seva.texter.presenter.source.LocationSource
import pl.org.seva.texter.presenter.utils.Constants
import pl.org.seva.texter.view.preference.HomeLocationPreference

class HomeLocationActivity : AppCompatActivity() {

    @Inject
    lateinit var locationSource: LocationSource

    private var locationChangedSubscription = Disposables.empty()

    private var lat: Double = 0.0
    private var lon: Double = 0.0
    private var toastShown: Boolean = false
    private var zoom: Float = 0.0f
    private var isCurrentLocationAvailable: Boolean = false

    private var map: GoogleMap? = null
    private var useCurrentButton: Button? = null
    private var animateCamera = true
    private var mapContainerId: Int = 0
    private var mapFragment: MapFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        savedInstanceState?.let {
            val myState = it.get(SAVED_STATE) as HomeLocationActivity.SavedState
            lat = myState.lat
            lon = myState.lon
            toastShown = myState.toastShown
            zoom = myState.zoom
            animateCamera = false
        }

        val graph = (application as TexterApplication).graph
        graph.inject(this)

        val value = persistedString
        lat = HomeLocationPreference.parseLatitude(value)
        lon = HomeLocationPreference.parseLongitude(value)

        setContentView(R.layout.activity_home_location)

        mapContainerId = findViewById(R.id.map_container).id
        MapsInitializer.initialize(this)
        useCurrentButton = findViewById(R.id.current_location_button) as Button

        val locationPermitted = ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        val locationAvailable = locationPermitted && locationSource.isLocationAvailable
        useCurrentButton?.isEnabled = locationAvailable
        useCurrentButton?.setOnClickListener { onUseCurrentLocationButtonClicked() }
        if (!toastShown) {
            Toast.makeText(
                    this,
                    R.string.long_press,
                    Toast.LENGTH_SHORT).show()
            toastShown = true
        }
        locationChangedSubscription = locationSource.addLocationChangedListener { onLocationChanged() }

        if (!locationPermitted) {
            useCurrentButton!!.isEnabled = false
        }
    }

    override fun onResume() {
        super.onResume()
        val fm = fragmentManager
        mapFragment = fm.findFragmentByTag(MAP_TAG_HOME_LOCATION) as MapFragment?
        mapFragment?:let {
            mapFragment = MapFragment()
            fm.beginTransaction().add(mapContainerId, mapFragment, MAP_TAG_HOME_LOCATION).commit()
        }
        mapFragment!!.getMapAsync({ this.onMapReady(it) })
    }

    private fun onMapReady(googleMap: GoogleMap) {
        map = googleMap
        if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            map!!.isMyLocationEnabled = true
        }
        zoom = PreferenceManager.getDefaultSharedPreferences(this).getFloat(ZOOM_PROPERTY_NAME, ZOOM_DEFAULT_VALUE)

        updateMarker()
        val cameraPosition = CameraPosition.Builder()
                .target(LatLng(lat, lon)).zoom(zoom).build()
        if (animateCamera) {
            map!!.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
        } else {
            map!!.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
        }
        animateCamera = false

        map!!.setOnMapLongClickListener({ this.onMapLongClick(it) })
        map!!.setOnCameraIdleListener({ this.onCameraIdle() })
    }

    override fun onBackPressed() {
        locationChangedSubscription.dispose()
        toastShown = false

        persistString(latLngToString())
        PreferenceManager.getDefaultSharedPreferences(this).edit().putFloat(ZOOM_PROPERTY_NAME, zoom).apply()
        locationSource.onHomeLocationChanged()

        mapFragment?.let {
            // Without enclosing in the if, throws:
            // java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
            fragmentManager.beginTransaction().remove(it).commit()
        }
        super.onBackPressed()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        val myState = HomeLocationActivity.SavedState()
        myState.lat = lat
        myState.lon = lon
        myState.toastShown = toastShown
        myState.zoom = zoom
        outState.putParcelable(SAVED_STATE, myState)

        mapFragment?.let {
            fragmentManager.beginTransaction().remove(it).commit()
            mapFragment = null
        }
        super.onSaveInstanceState(outState)
    }

    fun latLngToString(): String {
        return HomeLocationPreference.toString(lat, lon)
    }

    private fun persistString(`val`: String) {
        PreferenceManager.getDefaultSharedPreferences(this).edit()
                .putString(SettingsActivity.HOME_LOCATION, `val`).apply()
    }

    private val persistedString: String
        get() = PreferenceManager.getDefaultSharedPreferences(this).getString(SettingsActivity.HOME_LOCATION, Constants.DEFAULT_HOME_LOCATION)

    private fun updateMarker() {
        map!!.clear()
        val marker = MarkerOptions().position(
                LatLng(lat, lon)).title(getString(R.string.home))
        marker.icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_ROSE))
        map!!.addMarker(marker)
    }

    private fun onLocationChanged() {
        if (map == null || useCurrentButton == null) {
            return
        }
        isCurrentLocationAvailable = true
        useCurrentButton!!.isEnabled = true
    }


    private fun onMapLongClick(latLng: LatLng) {
        lat = latLng.latitude
        lon = latLng.longitude
        updateMarker()
        if (isCurrentLocationAvailable) {
            useCurrentButton!!.isEnabled = true
        }
    }

    private fun onCameraIdle() {
        zoom = map!!.cameraPosition.zoom
    }

    fun onUseCurrentLocationButtonClicked() {
        useCurrentButton!!.isEnabled = false
        val loc = locationSource.latLng
        loc?.let {
            lat = it.latitude
            lon = it.longitude
            updateMarker()
            val cameraPosition = CameraPosition.Builder()
                    .target(LatLng(lat, lon)).zoom(zoom).build()
            map!!.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
        }
    }

    private class SavedState internal constructor() : Parcelable {
        var lat: Double = 0.0
        var lon: Double = 0.0
        var toastShown: Boolean = false
        var zoom: Float = 0.0f

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeDouble(lat)
            dest.writeDouble(lon)
            dest.writeInt(if (toastShown) 1 else 0)
            dest.writeFloat(zoom)
        }
    }

    companion object {
        private val MAP_TAG_HOME_LOCATION = "MAP_HOME_LOCATIOT"

        private val SAVED_STATE = "saved_state"
        private val ZOOM_PROPERTY_NAME = "map_preference_gui_zoom"
        private val ZOOM_DEFAULT_VALUE = 7.5f
    }
}
